src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +221,122 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
+  (rewardsByAccountSpecifier, validatorAddress, denom): string => {
     if (!rewardsByAccountSpecifier.length) return ''
 
     const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byvalidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byvalidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const selectStakingOpportunityDataByDenom = createDeepEqualOutputSelector(
+  selectAllDelegationsCryptoAmountByDenom,
+  selectAllUnbondingsEntriesByDenom,
+  selectAllRewardsByDenom,
+  selectAllValidators,
+  (
+    allDelegationsAmount,
+    allUndelegationsEntries,
+    allRewards,
+    allValidators
+  ): StakingOpportunity[] => {
+    const result = Object.entries(allValidators).map(([validatorAddress, { apr, moniker }]) => {
+      const delegationsAmount = allDelegationsAmount[validatorAddress] ?? '0'
+      const undelegationsAmount = get(
+        allUndelegationsEntries,
+        validatorAddress,
+        [] as chainAdapters.cosmos.UndelegationEntry[]
+      )
+        .reduce((acc: BigNumber, undelegationEntry: chainAdapters.cosmos.UndelegationEntry) => {
+          acc = acc.plus(bnOrZero(undelegationEntry.amount))
+          return acc
+        }, bnOrZero(0))
+        .toString()
+
+      const rewards = get(allRewards, validatorAddress, [] as chainAdapters.cosmos.Reward[])
+        .reduce((acc: BigNumber, rewardEntry: chainAdapters.cosmos.Reward) => {
+          acc = acc.plus(bnOrZero(rewardEntry.amount))
+          return acc
+        }, bnOrZero(0))
+        .toString()
+
+      const cryptoAmount = bnOrZero(delegationsAmount)
+        .plus(bnOrZero(undelegationsAmount))
+        .toString()
+
+      return {
+        address: validatorAddress,
+        apr,
+        moniker,
+        cryptoAmount,
+        rewards
+      }
+    })
+
+    return result.filter(x => x.cryptoAmount !== '0' || x.rewards !== '0')
```

Comment:
[gomesalexandre](https://github.com/gomesalexandre) [yesterday](https://github.com/shapeshift/web/pull/1356/files#r840065724)
The filtering and parsing could be brought to one iteration, making this go from O(2n) to O(n).
Although at this point it might just be a better effort to go with smaller selectors and make each of these more performant, you could leverage using createSelector in them instead of createDeepOutputEqualSelector if they output primitives.

Reply:
Split into multiple memoize functions but didn't make it only one iteration yet, I would be interested to see how you would do this

Reply:
@NeOMakinG while I was at it, I replaced the map of this selector with a reduce, hence doing the filtering and parsing in one loop. Hope it helped.